backend/vertical-healthcare_ch/forum_datenaustausch_ch.invoice_450_request/src/main/java/de/metas/vertical/healthcare_ch/forum_datenaustausch_ch/invoice_450/Invoice450FromCrossVersionModelTool.java

```diff
@@ -1163,69 +1197,230 @@ private ServicesType createServicesType(@NonNull final List<XmlService> services
+		return xtraDrugType;
+	}
+
+	private ServiceType createRecordServiceType_XmlRecordDrg(final XmlRecordDrg xRecord)
+	{
+		final ServiceType serviceType = jaxbRequestObjectFactory.createServiceType();
+
+		setServiceTypeFields(serviceType, xRecord.getRecordService());
+		serviceType.setTariffType(xRecord.getTariffType());
+		serviceType.setXtraDrg(createXtraDrgType(xRecord));
+		return serviceType;
+	}
+
+	private XtraDRGType createXtraDrgType(final XmlRecordDrg xRecord)
+	{
+		final XtraDRGType xtraDRGType = jaxbRequestObjectFactory.createXtraDRGType();
+		xtraDRGType.setCostFraction(xRecord.getCostFraction());
+		return xtraDRGType;
+	}
+
+	private void setServiceTypeFields(
+			@NonNull final ServiceType recordServiceType,
+			@NonNull final XmlRecordService recordService)
+	{
+		recordServiceType.setRecordId(recordService.getRecordId());
+		recordServiceType.setCode(recordService.getCode());
+		recordServiceType.setRefCode(recordService.getRefCode());
+		recordServiceType.setName(recordService.getName());
+		recordServiceType.setSession(recordService.getSession());
+		recordServiceType.setQuantity(recordService.getQuantity());
+		recordServiceType.setDateBegin(recordService.getDateBegin());
+		recordServiceType.setDateEnd(recordService.getDateEnd());
+		recordServiceType.setProviderId(getEanPartyId(recordService.getProviderId()));
+		recordServiceType.setResponsibleId(getEanPartyId(recordService.getResponsibleId()));
+		recordServiceType.setUnit(recordService.getUnit());
+		recordServiceType.setUnitFactor(recordService.getUnitFactor());
+		recordServiceType.setExternalFactor(recordService.getExternalFactor());
+		recordServiceType.setAmount(recordService.getAmount());
+		recordServiceType.setVatRate(recordService.getVatRate());
+		recordServiceType.setObligation(recordService.getObligation());
+		recordServiceType.setSectionCode(recordService.getSectionCode());
+		recordServiceType.setRemark(recordService.getRemark());
+		recordServiceType.setServiceAttributes(recordService.getServiceAttributes());
+	}
+
+	/**
+	 * In version 4.4 both the {@code providerId} and the {@code responsibleId} of a service could be specified via either {@code eanPartyType} or {@code zsrPartyType}.
+	 * In version 4.5 only {@code eanPartyType} is supported. Attempting to figure this value out from other places where both zsrPartyType and eanPartyType are defined, namely:
+	 * <ul>
+	 *     	<li>{@code billerAddressType}</li>
+	 *     	<li>{@code providerAddressType}</li>
+	 *     	<li>{@code referrerAddressType}</li>
+	 * </ul>
+	 *
+	 * @param zsrOrEanPartyId
+	 * @return
+	 */
+	private String getEanPartyId(@NonNull final String zsrOrEanPartyId)
+	{
+		if (zsrToEanPartyMap.containsKey(zsrOrEanPartyId))
+		{
+			return zsrToEanPartyMap.get(zsrOrEanPartyId);
+		}
+		else if (!StringUtils.isNumber(zsrOrEanPartyId))
+		{
+			Check.fail("Invalid ean: {}", zsrOrEanPartyId);
+		}
+		return zsrOrEanPartyId;
+	}
```

Comment:
This is honestly a shot in the dark, no idea how other to solve this problem. If we're translating from 4.4 and we get a value as zsr that we didn't previously pick up along with an EanParty (to know where to map it to), we're dead in the water and can't convert from 4.4

Reply:
Suggestions:
- explain the problem to the customer. Maybe they can ensure on their side to always deliver just EANs.
- here are some screenshots you might use:

![image](https://user-images.githubusercontent.com/1264789/146871163-5c950a66-aff7-4194-96e7-b0eae2ca6100.png)
![image](https://user-images.githubusercontent.com/1264789/146871170-654761ae-427c-4d0c-a41e-a29ee19648f9.png)

**If no**
- (do C&E first pls)
- Get the windows-help-file https://www.forum-datenaustausch.ch/fileadmin/fda_documents/4.5/Rechnungen/XML/generalInvoiceRequest450.chm_210304.zip and check out what those EANs are about
  - note that in order to read it on a sane windows system, you need to "unlock" it in the properties: 
![image](https://user-images.githubusercontent.com/1264789/146871141-29d8e3d9-fd7a-43da-bd00-ae4fe8362fd7.png)
  - spoiler: I took a quick look, but could not see whether or not we can generally expect that the zrs/ean combos show up earlier in the XML.
- Extend `XmlRecordService` to have either a dedicated ZSR and EAN field, or to have just an additional EAN field in addition to the existing "both"-field.
- Extend `XmlService.ServiceModWithSelector` so they can be set within the X-model by metasfresh
- Implement a ZSR => EAN mapping table and window in metasfresh that the user needs to maintain
  - Use the descriptions and terminology from the `chm`-file for the fields' names and descriptions.
- If there is a ZSR, then set it's respective EAN from this table to the x-model.